### PR TITLE
Pymeep 1.17.1 (attempting to fix conflicts in previous PRs)

### PR DESCRIPTION
**Pull request by bdice**
I'm using this PR to draft some changes that can run on CI (my local machine won't run the Docker build). I'm going to try and fix the issues in the open PRs with dependency conflicts and will open this PR once I have more information that might point towards a solution.

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by bdice** [comment]
@oskooi Thanks for your assistance with the PRs I made to meep dependencies. I'm going to try and debug this further tomorrow. I may need to re-render/update the other dependencies (libgdsii, maybe) until the dependency conflicts are solved. If that doesn't fix it, I am going to systematically eliminate non-critical dependencies and disable MPI builds until I get something that passes CI, then add back that functionality once I can find the source of the problem. It's quite difficult to read conda's conflict outputs in the logs, unless I'm just missing something obvious.

**Comment by oskooi** [comment]
@bdice, many thanks for your help updating the PyMeep 1.17.1 package.

Note that the [package for Linux](https://anaconda.org/conda-forge/pymeep) has already been updated to version 1.17.1. This happened automatically even though #54 and #55 were never merged. The package for macOS is still only at 1.16.1, however.

**Comment by bdice** [comment]
@oskooi Good point, I noticed that when looking at the branch in #54. The specific upgrade I need is Python 3.9 support, and ideally MPI builds. This PR won’t add the Python 3.9 support but a working current version will allow us to trigger the migration again and get Python 3.9 added.

**Comment by bdice** [comment]
@oskooi The tests are passing now. I removed the following "extras" and I would like your help to identify whether these are actually safe to remove. I think it's a good idea to define only the minimal dependencies, especially if the application can raise user-friendly messages to indicate features that require additional dependencies.

```yaml
    # Extras typically used at runtime
    #- h5utils * {{ mpi_prefix }}_*
    #- matplotlib-base >=2.2
    #- scipy
    #- mayavi
    #- autograd
    #- ffmpeg
    #- zlib
    #- libopenblas
```

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,12 +1,15 @@
 {% set name = "pymeep" %}
-{% set version = "1.16.1" %}
-{% set sha256 = "b0dc3c0090a6562eb67ad1dd986a3347dc188c0816dce3c5ec1e8870428d41d3" %}
+{% set version = "1.17.1" %}
+{% set sha256 = "cea11aafe8be1d7ab762324d2dc7d9f4a9b713cee87f9722ced2c5cc30f53a99" %}
 {% set buildnumber = 0 %}
```
Comment (bdice) [inline recipe/meta.yaml:4]:
```suggestion
{% set buildnumber = 1 %}
```

**Comment by oskooi** [comment]
It seems that removing the "extra" package dependencies causes only *some* of the tests in the `make check` suite to run while others are excluded. For example, one of the tests ([adjoint_solver.py](https://github.com/NanoComp/meep/blob/master/python/tests/adjoint_solver.py)) uses `autograd` but this is not shown in the Travis logs. In fact, only 19 out of a total of 47 tests in the `python/tests` subredictory actually ran (all of them successfully). Also, the [tutorials in the user manual](https://meep.readthedocs.io/en/latest/Python_Tutorials/Basics/) require the "extra" dependencies for post-processing the simulation output.

Thus, it would be good to include all the dependencies in the original recipe. Note that one package conflict involves BLAS and NumPy as described in [#38 (comment)](https://github.com/conda-forge/pymeep-feedstock/issues/38#issuecomment-670574988).

**Comment by bdice** [comment]
@oskooi I would propose adding some of these extra runtime dependencies to the “testing dependencies” list. That way, the package will be tested with all of them, but they won’t be installed unless needed by the user. For example, I won’t be using any of these features and thus don’t need the packages installed. Additionally, putting all of the visualization tools like mayavi as runtime dependencies means that meep cannot be built for ppc64le or ARM (those packages were causing conflicts on my previous PR).

For the tutorials, you could either list the packages explicitly in the tutorial, or create a meta-package that includes all of them and instruct users to install that. What do you think? If you support these ideas, I can help implement them.

**Comment by oskooi** [comment]
Thanks for your suggestion. Let's go with putting these extra packages into a "testing dependencies" list and also create a "meta-package" for users to install on their own, as necessary.

**Comment by bdice** [comment]
Great. I will add the testing dependencies first, and then work on the metapackage. What name would you suggest for the metapackage, `meep-tutorial` or `meep-extras` or something else?

**Comment by oskooi** [comment]
I think `meep-extras` is an appropriate name.

**Comment by bdice** [comment]
## Testing dependencies
It looks like these packages are used in the tests, so I added them as testing dependencies in cf3ce6c45f5b742370f9b82f29fd171f456d1ae4.
- `autograd` -- used in `pytest/tests/adjoint_solver.py`
- `scipy` -- used in `pytest/tests/adjoint_solver.py`
- `matplotlib-base` -- used in `python/tests/visualization.py`
- `ffmpeg` -- used in `python/tests/visualization.py`

It looks like these packages would be suitable for the `meep-extras` metapackage.
- (all of the above testing dependencies)
- `mayavi` -- used for 3D rendering in the tutorial
- `h5utils * {{ mpi_prefix }}_*` -- needed for the `h5topng` command in the tutorial

I'm not sure what these packages are meant for, and clarification would be helpful if you happen to know more:
- `zlib` - is this used?
- `libopenblas` - I read the [linked comment above](https://github.com/conda-forge/pymeep-feedstock/issues/38#issuecomment-670574988), do you think this BLAS conflict is still an issue with the current conda-forge packages? I would have expected to see a failure by now if it is still a current problem. If not, let's just remove this dependency.
- `parallel` - I didn't see the GNU parallel utility used in the tutorial or tests. It is currently listed as a "testing dependency" but it might be safe to remove (haven't tried yet).

## Metapackage

I am going to use the [outputs section](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#package-outputs) of the recipe to define two packages named `pymeep` and `pymeep-extras`. Here's a [fairly complicated example](https://github.com/conda-forge/binutils-feedstock/blob/master/recipe/meta.yaml) that demonstrates how the syntax will look. I'm going to make that change after we finalize testing dependencies.

**Comment by oskooi** [comment]
BLAS and LAPACK are necessary for compiling Meep dependencies [MPB](https://github.com/NanoComp/mpb) and [harminv](https://github.com/NanoComp/harminv). It seems `libopenblas` was added in #44 to possibly force a NumPy dependency which is compatible with these two packages? 

`zlib` was added in #44 and also in conda-forge/mpb-feedstock#28. Not sure whether it can be removed.

`parallel` was added in the [initial feedstock commit](https://github.com/conda-forge/pymeep-feedstock/commit/b46adb9e89a99784b19119b696e9964ad8bad35c). Perhaps it is necessary and should be retained?

**Comment by bdice** [comment]
I confirmed that `parallel` is needed for running tests, and I will add it back to the testing dependencies.

I see the warnings that were mentioned in those previous PRs, so I will add back `zlib` and `libopenblas` (and `_openmp_mutex` on Linux) to the runtime dependencies. For future reference, the warnings look like this (CI build logs will eventually be purged and leave no clear evidence for why we did this):

```
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): lib/libmeep.so.23 found in this package
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libGDSII.so.0 found in conda-forge::libgdsii-0.21-h4bd325d_4
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libctlgeom.so.7 found in conda-forge::libctl-4.5.0-h7f98852_2
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libhdf5.so.103 found in conda-forge::hdf5-1.10.5-mpi_mpich_he1803ad_1014
WARNING (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libz.so.1 found in ['zlib']
WARNING (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): .. but ['zlib'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libgsl.so.25 found in conda-forge::gsl-2.6-he838d99_2
WARNING (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libopenblas.so.0 found in ['libopenblas']
WARNING (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): .. but ['libopenblas'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libmpb.so.1 found in conda-forge::mpb-1.11.1-mpi_mpich_h173c125_1
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libharminv.so.3 found in conda-forge::harminv-1.4.1-hbf3b5f1_1007
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libfftw3.so.3 found in conda-forge::fftw-3.3.9-mpi_mpich_h40e55c9_0
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libgfortran.so.5 found in conda-forge::libgfortran5-9.3.0-hff62375_18
WARNING (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libgomp.so.1 found in ['_openmp_mutex']
WARNING (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): .. but ['_openmp_mutex'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libquadmath.so.0 found in conda-forge::libgcc-ng-9.3.0-h2828fa1_18
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libpthread.so.0 found in CDT/compiler package conda-forge::sysroot_linux-64-2.12-h77966d4_13
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libmpicxx.so.12 found in conda-forge::mpich-3.4.1-h846660c_2
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libmpi.so.12 found in conda-forge::mpich-3.4.1-h846660c_2
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libstdc++.so.6 found in conda-forge::libstdcxx-ng-9.3.0-h6de172a_18
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libm.so.6 found in CDT/compiler package conda-forge::sysroot_linux-64-2.12-h77966d4_13
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO x86_64-conda-linux-gnu/sysroot/lib64/libc.so.6 found in CDT/compiler package conda-forge::sysroot_linux-64-2.12-h77966d4_13
   INFO (pymeep,lib/python3.6/site-packages/meep/_meep.so.23.0.1): Needed DSO lib/libgcc_s.so.1 found in conda-forge::libgcc-ng-9.3.0-h2828fa1_18
```

**Comment by bdice** [comment]
I'm stopping on this for the night -- I'll revisit tomorrow and hopefully we can merge then. I'm working with the pymeep-extras metapackage now, and trying to find the right way to write it. I am following this example: https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html#referencing-subpackages

**Comment by oskooi** [comment]
Ok, thanks for all your help thus far.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
+{% set mpi = mpi or "nompi" %}
 {% if mpi == "nompi" %}
 # prioritize nompi via build number
 {% set buildnumber = buildnumber + 100 %}
+{% set mpi_prefix = "nompi" %}
+{% else %}
+{% set mpi_prefix = "mpi_" + mpi %}
 {% endif %}
 
 package:
```
Comment (jaimergp) [inline recipe/meta.yaml:15]: Try using a different global name so it does not collide with the subpackages names. It has led to surprising behavior in the past, and might help here. Some naming conventions people usually follow is `name: {{ name }}-split`.

**Comment by jaimergp** [comment]
@bdice Try playing with the global package name so it does not conflict with subpackages. Also, I don't think you need to run the test suite in the `*-extras` package; checking that the test environment is solvable should be enough. 

Also, I am not sure, but maybe `*-extras` can be `noarch`? Somebody from Core should confirm though.

**Comment by jaimergp** [comment]
Oh, and maybe you need to use a different script name instead of `build.sh` / `bld.bat`, and refer to the new name in the `outputs[0].build.script` entry. This can only be a filename so far, due to some conda build limitations, I am afraid.

**Comment by bdice** [comment]
@jaimergp I'm _super_ appreciative of your suggestions, all of those made sense! 🙇🙇🙇 It's almost working, there is some HDF5 conflict now that I wasn't seeing before the feedstock split...

**Comment by bdice** [comment]
I have a (hopefully correct) diagnosis of the last remaining problem.

During deployment of the last mpb builds, the linux64 mpich build randomly failed (http connection failure).
https://github.com/conda-forge/mpb-feedstock/runs/1958833526

This led to the failure of the linux64 mpich build of pymeep to fail, because the pinnings for hdf5 are inconsistent.
```
2021-02-25T23:57:33.5366212Z Package hdf5 conflicts for:
2021-02-25T23:57:33.5367021Z hdf5[version='*,1.10.6.*',build=mpi_mpich_*]
2021-02-25T23:57:33.5367949Z mpb[version='>=1.11',build=mpi_mpich_*] -> hdf5[version='1.10.5.*|>=1.10.5,<1.10.6.0a0|>=1.10.5,<1.10.6.0a0',build=mpi_mpich_*]
```

We can fix this by re-deploying mpb, I think. I'm going to make a PR that re-renders MPB and it can be merged once tests pass. Then this PR's CI can be re-triggered and everything should be green. 🤞

**Comment by bdice** [comment]
I think I triggered CI too early. It takes a little while (30 min?) for packages to be available, and it looks like the old mpb package was found again on linux64 + mpich. I will wait a little longer and then trigger CI again.

**Comment by bdice** [comment]
@oskooi I think CI will pass now. This PR is ready for review! 👍

**Comment by oskooi** [comment]
@bdice, nice work.

**Comment by jaimergp** [comment]
@bdice Glad to see you got this working! 🥳

**Pull request merged**